### ISSUE TITLE
Encoding problem in cover letters

### ISSUE THREAD

**Issue by philmd** [stefanha/git-publish#88]
See for example this cover letter:
https://lists.gnu.org/archive/html/qemu-devel/2020-06/msg01296.html
```
  io_uring: retry io_uring_submit() if it fails with errno=3DEINTR
  io_uring: use io_uring_cq_ready() to check for ready cqes
```
Patch content and subject are however correct:
https://lists.gnu.org/archive/html/qemu-devel/2020-06/msg01297.html
```
[PULL 01/11] io_uring: retry io_uring_submit() if it fails with errno=EINTR
```

**Comment by stefanha** [comment]
Hi @philmd, can you confirm which version of git-publish you are using?

**Comment by philmd** [comment]
@stefanha git-publish 1.6.1

**Comment by stefanha** [comment]
@philmd Thanks for confirming. In that case it will be necessary to reproduce and debug this since it hasn't been fixed yet. Let's keep this GitHub Issue open until someone has time to investigate.

**Comment by stefano-garzarella** [comment]
I had a similar issue here: https://lists.gnu.org/archive/html/qemu-devel/2020-08/msg04373.html

I looked in the 0000-cover-letter.patch and before it was sent, I found `Content-Transfer-Encoding: quoted-printable`:
```
From: Stefano Garzarella <sgarzare@redhat.com>
Date: Wed, 19 Aug 2020 16:34:46 +0200
Subject: [PATCH v2 0/3] vhost-vsock: force virtio version 1
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: quoted-printable
MIME-Version: 1.0
```

So I forced 8bit CTE in this way and it seems to work :

```patch
diff --git a/git-publish b/git-publish
index 1920979..868a27c 100755
--- a/git-publish
+++ b/git-publish
@@ -846,7 +846,7 @@ branch.%s.pushRemote is set appropriately?  (Override with --no-url-check)''' %
 
                 blurb = os.linesep.join(message[2:])
                 body = msg.get_content().replace('*** BLURB HERE ***', blurb)
-                msg.set_content(body)
+                msg.set_content(body, charset='utf-8', cte='8bit')
 
                 open(cover_letter_path, 'wb').write(msg.as_bytes())
             patches = sorted(glob.glob(os.path.join(tmpdir, '*')))
```

Do you think that's an acceptable solution?

**Comment by stefanha** [comment]
On Wed, Aug 19, 2020 at 3:41 PM Stefano Garzarella <notifications@github.com>
wrote:

> I had a similar issue here:
> https://lists.gnu.org/archive/html/qemu-devel/2020-08/msg04373.html
>
> I looked in the 0000-cover-letter.patch and before it was sent, I found Content-Transfer-Encoding:
> quoted-printable:
>
> From: Stefano Garzarella <sgarzare@redhat.com>
> Date: Wed, 19 Aug 2020 16:34:46 +0200
> Subject: [PATCH v2 0/3] vhost-vsock: force virtio version 1
> Content-Type: text/plain; charset="utf-8"
> Content-Transfer-Encoding: quoted-printable
> MIME-Version: 1.0
>
> So I forced 8bit CTE in this way and it seems to work :
>
> diff --git a/git-publish b/git-publish
> index 1920979..868a27c 100755--- a/git-publish+++ b/git-publish@@ -846,7 +846,7 @@ branch.%s.pushRemote is set appropriately?  (Override with --no-url-check)''' %
>
>                  blurb = os.linesep.join(message[2:])
>                  body = msg.get_content().replace('*** BLURB HERE ***', blurb)-                msg.set_content(body)+                msg.set_content(body, charset='utf-8', cte='8bit')
>
>                  open(cover_letter_path, 'wb').write(msg.as_bytes())
>              patches = sorted(glob.glob(os.path.join(tmpdir, '*')))
>
> Do you think that's an acceptable solution?
>
That's great. Before applying that fix I would like to understand why it
works.

The documentation says charset='utf-8' is the default:
https://docs.python.org/3/library/email.contentmanager.html#email.contentmanager.raw_data_manager

quoted-printable is a binary->ascii encoding. It should handle any 8-bit
value so it's unclear to me why there would be an encoding problem.

One guess is that email.message isn't updating the headers when
set_content() is called?

Can you explain why the fix works?

**Comment by stefano-garzarella** [comment]
The issue seems related to 'sendemail.transferEncoding' configuration in my ~/.gitconfig

I set it to '8bit', so when the cover letter is encoded in quoted-printable `git send-email` changes the `Content-Transfer-Encoding` in the header, without re-encoding the body.

I set it to 'auto' and the `Content-Transfer-Encoding` is right, but the body seems "double" encoded...

```
Recenlty changes in QEMU 5.1 requires to set 'disable-legacy=3D3Don'
```

So I think the patch proposed works only for my case, I need to understand better why `git send-email` change the `Content-Transfer-Encoding`.

Maybe it is expecting that the cover-letter is not encoded.
I'll get a better look in the next days.
